File: Euler55/src/main.rs
use std::collections::VecDeque;

fn main() {
    let amount = (1..10000).fold(0, |acc, n| if is_lychrel(n) { acc + 1 } else { acc });
    println!("Amount is {}", amount);
}

fn is_lychrel(mut n: u64) -> bool {
    const MAX_ITERATIONS: u64 = 50;
    let mut counter = MAX_ITERATIONS;
    while counter > 0 {
        let reverse = reverse_number(n);
        n += reverse;
        counter -= 1;
        if n == reverse_number(n) {
            return false;
        }
    }
    true
}

fn reverse_number(mut n: u64) -> u64 {
    let mut digits = VecDeque::new();
    while n > 0 {
        digits.push_front(n % 10);
        n /= 10;
    }
    let mut out: u64 = 0;
    loop {
        let next = digits.pop_back();
        match next {
            Some(num) => { 
                let ten_x = out.checked_mul(10);
                match ten_x {
                    Some(multiplied) => { out = multiplied + num; }
                    None => { return 0; }
                }
            }
            None => { break; }
        }
    }
    out
}

#[test]
fn test_is_lychrel() {
    assert!(!is_lychrel(47));
    assert!(!is_lychrel(349));
    assert!(is_lychrel(196));
    assert!(is_lychrel(4994));
}

#[test]
fn test_reverse_number() {
    assert_eq!(reverse_number(12345), 54321);
    assert_eq!(reverse_number(555555), 555555);
    assert_eq!(reverse_number(0), 0);
    assert_eq!(reverse_number(1), 1);
}

File: Euler55/main.py
from functools import reduce
def reverse_number(n):
    out = 0
    while n > 0:
        out *= 10
        out += n % 10
        n //= 10
    return out

def is_lychrel(n):
    orig = n
    for _ in range(50):
        n += reverse_number(n)
        if n == reverse_number(n):
            return False
    return True

assert reverse_number(12345) == 54321
assert reverse_number(1111) == 1111
assert reverse_number(0) == 0

assert not is_lychrel(47)
assert is_lychrel(196)
assert not is_lychrel(349)
assert is_lychrel(4994)

print(reduce(lambda x, y: x + (1 if is_lychrel(y) else 0), range(10, 10000), 0))
